File: powerups.py
import random
import pygame
from config import (
    WINDOW_WIDTH, 
    WINDOW_HEIGHT,
    RETRO_CYAN,
    RETRO_YELLOW,
    RETRO_GREEN,
    RETRO_MAGENTA,
    RETRO_SURFACE,
    RETRO_WHITE
)

class RetroPowerUp:
    def __init__(self, type_name):
        self.type = type_name
        self.x = random.randint(50, WINDOW_WIDTH - 50)
        self.y = -30
        self.speed = 0.8
        self.is_active = True
        self.size = 16
        
        # Retro colors and simple geometric symbols
        self.colors = {
            "freeze": RETRO_CYAN,
            "clear": RETRO_YELLOW,
            "life": RETRO_GREEN,
            "shield": RETRO_MAGENTA
        }
        
        # Simple geometric shapes instead of emojis
        self.shapes = {
            "freeze": "diamond",
            "clear": "star",
            "life": "heart",
            "shield": "square"
        }
        
    def update(self):
        self.y += self.speed
        return self.y < WINDOW_HEIGHT + 30
        
    def draw(self, screen, font):
        if not self.is_active:
            return
            
        color = self.colors[self.type]
        shape = self.shapes[self.type]
        
        # Draw different retro shapes
        if shape == "diamond":
            # Diamond shape for freeze
            points = [
                (self.x, self.y - self.size),
                (self.x + self.size, self.y),
                (self.x, self.y + self.size),
                (self.x - self.size, self.y)
            ]
            pygame.draw.polygon(screen, color, points)
            pygame.draw.polygon(screen, RETRO_WHITE, points, 2)
            
        elif shape == "star":
            # Star shape for clear (simplified as plus)
            pygame.draw.rect(screen, color, (self.x - self.size, self.y - 3, self.size * 2, 6))
            pygame.draw.rect(screen, color, (self.x - 3, self.y - self.size, 6, self.size * 2))
            pygame.draw.rect(screen, RETRO_WHITE, (self.x - self.size, self.y - 3, self.size * 2, 6), 1)
            pygame.draw.rect(screen, RETRO_WHITE, (self.x - 3, self.y - self.size, 6, self.size * 2), 1)
            
        elif shape == "heart":
            # Heart shape for life (simplified as circle)
            pygame.draw.circle(screen, color, (int(self.x), int(self.y)), self.size)
            pygame.draw.circle(screen, RETRO_WHITE, (int(self.x), int(self.y)), self.size, 2)
            
        elif shape == "square":
            # Square shape for shield
            rect = pygame.Rect(self.x - self.size, self.y - self.size, self.size * 2, self.size * 2)
            pygame.draw.rect(screen, color, rect)
            pygame.draw.rect(screen, RETRO_WHITE, rect, 2)
        
    def is_collected(self, word_x, word_y):
        """Check if a typed word is close enough to collect this power-up"""
        distance = ((self.x - word_x) ** 2 + (self.y - word_y) ** 2) ** 0.5
        return distance < self.size * 2

class PowerUpManager:
    def __init__(self):
        self.powerups = []
        self.font = pygame.font.Font(None, 20)
        self.ui_font = pygame.font.Font(None, 16)
        self.spawn_chance = 0.01  # 1% chance per update
        self.active_effects = {
            "freeze": 0,
            "shield": 0
        }
        
    def update(self):
        # Update existing powerups
        self.powerups = [p for p in self.powerups if p.update() and p.is_active]
        
        # Update active effect timers
        for effect in list(self.active_effects.keys()):
            if self.active_effects[effect] > 0:
                self.active_effects[effect] -= 1
                
    def draw(self, screen):
        # Draw powerups
        for powerup in self.powerups:
            powerup.draw(screen, self.font)
            
        # Draw retro active effects panel
        if any(time > 0 for time in self.active_effects.values()):
            panel_width = 150
            panel_height = 60
            panel_x = WINDOW_WIDTH - panel_width - 10
            panel_y = 150
            
            # Create effects panel
            panel_rect = pygame.Rect(panel_x, panel_y, panel_width, panel_height)
            pygame.draw.rect(screen, RETRO_SURFACE, panel_rect)
            pygame.draw.rect(screen, RETRO_WHITE, panel_rect, 1)
            
            # Draw active effects
            y_offset = 10
            for effect, time in self.active_effects.items():
                if time > 0:
                    seconds = time // 60 + 1
                    effect_text = f"{effect.upper()}: {seconds}S"
                    
                    color = RETRO_CYAN if effect == "freeze" else RETRO_MAGENTA
                    
                    text_surface = self.ui_font.render(effect_text, False, color)
                    screen.blit(text_surface, (panel_x + 5, panel_y + y_offset))
                    
                    # Simple progress bar
                    max_time = 300 if effect == "freeze" else 600
                    progress = time / max_time
                    bar_width = 100
                    bar_height = 4
                    bar_x = panel_x + 5
                    bar_y = panel_y + y_offset + 15
                    
                    # Background
                    pygame.draw.rect(screen, (50, 50, 50), (bar_x, bar_y, bar_width, bar_height))
                    # Progress
                    pygame.draw.rect(screen, color, (bar_x, bar_y, int(bar_width * progress), bar_height))
                    
                    y_offset += 25
                    
    def try_spawn(self):
        """Try to spawn a new powerup based on chance"""
        if random.random() < self.spawn_chance and len(self.powerups) < 2:
            powerup_type = random.choice(["freeze", "clear", "life", "shield"])
            self.powerups.append(RetroPowerUp(powerup_type))
            
    def check_collection(self, word_x, word_y):
        """Check if any powerup is collected by a correctly typed word"""
        for powerup in self.powerups:
            if powerup.is_active and powerup.is_collected(word_x, word_y):
                powerup.is_active = False
                return powerup.type
        return None
        
    def activate_effect(self, effect_type, duration=300):
        """Activate a powerup effect"""
        if effect_type in self.active_effects:
            self.active_effects[effect_type] = duration
            
    def is_effect_active(self, effect_type):
        """Check if an effect is currently active"""
        return self.active_effects.get(effect_type, 0) > 0
        
    def clear_powerups(self):
        """Clear all powerups from screen"""
        self.powerups.clear()
        self.active_effects = {key: 0 for key in self.active_effects}

File: js/powerups.js
// Powerup system for retro game
class RetroPowerUp {
  constructor(type) {
    this.type = type
    this.x = Math.random() * (640 - 100) + 50
    this.y = -30
    this.speed = 0.8
    this.isActive = true
    this.size = 16

    // Retro colors and shapes
    this.colors = {
      freeze: "#55ffff",
      clear: "#ffff55",
      life: "#55ff55",
      shield: "#ff55ff",
    }

    this.shapes = {
      freeze: "diamond",
      clear: "star",
      life: "heart",
      shield: "square",
    }
  }

  update() {
    this.y += this.speed
    return this.y < 510
  }

  draw(ctx) {
    if (!this.isActive) return

    const color = this.colors[this.type]
    const shape = this.shapes[this.type]

    ctx.fillStyle = color
    ctx.strokeStyle = "#ffffff"
    ctx.lineWidth = 2

    switch (shape) {
      case "diamond":
        // Diamond shape for freeze
        ctx.beginPath()
        ctx.moveTo(this.x, this.y - this.size)
        ctx.lineTo(this.x + this.size, this.y)
        ctx.lineTo(this.x, this.y + this.size)
        ctx.lineTo(this.x - this.size, this.y)
        ctx.closePath()
        ctx.fill()
        ctx.stroke()
        break

      case "star":
        // Star shape for clear (simplified as plus)
        ctx.fillRect(this.x - this.size, this.y - 3, this.size * 2, 6)
        ctx.fillRect(this.x - 3, this.y - this.size, 6, this.size * 2)
        ctx.strokeRect(this.x - this.size, this.y - 3, this.size * 2, 6)
        ctx.strokeRect(this.x - 3, this.y - this.size, 6, this.size * 2)
        break

      case "heart":
        // Heart shape for life (simplified as circle)
        ctx.beginPath()
        ctx.arc(this.x, this.y, this.size, 0, Math.PI * 2)
        ctx.fill()
        ctx.stroke()
        break

      case "square":
        // Square shape for shield
        ctx.fillRect(this.x - this.size, this.y - this.size, this.size * 2, this.size * 2)
        ctx.strokeRect(this.x - this.size, this.y - this.size, this.size * 2, this.size * 2)
        break
    }
  }

  isCollected(wordX, wordY) {
    const distance = Math.sqrt((this.x - wordX) ** 2 + (this.y - wordY) ** 2)
    return distance < this.size * 2
  }
}

class PowerUpManager {
  constructor() {
    this.powerups = []
    this.spawnChance = 0.01 // 1% chance per update
    this.activeEffects = {
      freeze: 0,
      shield: 0,
    }
  }

  update() {
    // Update existing powerups
    this.powerups = this.powerups.filter((powerup) => powerup.update() && powerup.isActive)

    // Update active effect timers
    for (const effect in this.activeEffects) {
      if (this.activeEffects[effect] > 0) {
        this.activeEffects[effect]--
      }
    }
  }

  draw(ctx) {
    // Draw powerups
    this.powerups.forEach((powerup) => powerup.draw(ctx))

    // Draw active effects panel
    if (Object.values(this.activeEffects).some((time) => time > 0)) {
      const panelWidth = 150
      const panelHeight = 60
      const panelX = 640 - panelWidth - 10
      const panelY = 150

      // Panel background
      ctx.fillStyle = "#282850"
      ctx.fillRect(panelX, panelY, panelWidth, panelHeight)
      ctx.strokeStyle = "#ffffff"
      ctx.lineWidth = 1
      ctx.strokeRect(panelX, panelY, panelWidth, panelHeight)

      // Draw active effects
      let yOffset = 10
      for (const [effect, time] of Object.entries(this.activeEffects)) {
        if (time > 0) {
          const seconds = Math.floor(time / 60) + 1
          const effectText = `${effect.toUpperCase()}: ${seconds}S`

          const color = effect === "freeze" ? "#55ffff" : "#ff55ff"

          ctx.font = "8px monospace"
          ctx.fillStyle = color
          ctx.textAlign = "left"
          ctx.fillText(effectText, panelX + 5, panelY + yOffset + 8)

          // Progress bar
          const maxTime = effect === "freeze" ? 300 : 600
          const progress = time / maxTime
          const barWidth = 100
          const barHeight = 4
          const barX = panelX + 5
          const barY = panelY + yOffset + 15

          // Background
          ctx.fillStyle = "#323232"
          ctx.fillRect(barX, barY, barWidth, barHeight)

          // Progress
          ctx.fillStyle = color
          ctx.fillRect(barX, barY, barWidth * progress, barHeight)

          yOffset += 25
        }
      }
    }
  }

  trySpawn() {
    if (Math.random() < this.spawnChance && this.powerups.length < 2) {
      const types = ["freeze", "clear", "life", "shield"]
      const type = types[Math.floor(Math.random() * types.length)]
      this.powerups.push(new RetroPowerUp(type))
    }
  }

  checkCollection(wordX, wordY) {
    for (const powerup of this.powerups) {
      if (powerup.isActive && powerup.isCollected(wordX, wordY)) {
        powerup.isActive = false
        return powerup.type
      }
    }
    return null
  }

  activateEffect(effectType, duration = 300) {
    if (effectType in this.activeEffects) {
      this.activeEffects[effectType] = duration
    }
  }

  isEffectActive(effectType) {
    return this.activeEffects[effectType] > 0
  }

  clearPowerups() {
    this.powerups = []
    for (const effect in this.activeEffects) {
      this.activeEffects[effect] = 0
    }
  }
}
